File: public_html/cgi-bin/php/state-demo-2.php
<?php
declare(strict_types=1);

header('Content-Type: text/html; charset=utf-8');
header('X-Content-Type-Options: nosniff');

$sessionName = 'SID';
session_name($sessionName);

if (isset($_GET['sid']))  session_id(preg_replace('/[^a-zA-Z0-9,-]/', '', $_GET['sid']));
if (isset($_POST['sid'])) session_id(preg_replace('/[^a-zA-Z0-9,-]/', '', $_POST['sid']));

session_start([
  'cookie_httponly' => true,
  'cookie_secure'   => (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off'),
]);

$transport = $_GET['transport'] ?? 'cookie';
if (!in_array($transport, ['cookie','url','hidden'], true)) $transport = 'cookie';

if (isset($_POST['name'])) {
    $_SESSION['name'] = trim((string)$_POST['name']);
}

function link_to(string $path, array $params = []): string {
    return htmlspecialchars($path . (empty($params) ? '' : '?' . http_build_query($params)), ENT_QUOTES, 'UTF-8');
}

$paramsBack = ['transport' => $transport];
$paramsSelf = ['transport' => $transport];
if ($transport === 'url') {
    $paramsBack['sid'] = session_id();
    $paramsSelf['sid'] = session_id();
}

$name = $_SESSION['name'] ?? '';
$hello = $name !== '' ? "Hello, $name!" : 'No name stored yet.';
?>
<!doctype html>
<html>
<head>
<meta charset="utf-8">
<title>State Demo — Page 2 (PHP)</title>
<style>
  body{font-family:system-ui,Arial,sans-serif;margin:2rem}
  code{background:#f6f6f6; padding:.1rem .3rem; border-radius:4px}
  form{margin-top:1rem}
  button{padding:.45rem .8rem}
</style>
</head>
<body>
  <h1>State Demo — Page 2</h1>
  <p><b>Session ID:</b> <code><?= htmlspecialchars(session_id(), ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8') ?></code></p>
  <h2><?= htmlspecialchars($hello, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8') ?></h2>

  <p><a href="<?= link_to('/cgi-bin/php/state-demo-1.php', $paramsBack) ?>">Back to Page 1</a></p>

  <form method="post" action="<?= link_to('/cgi-bin/php/state-demo-2.php', $paramsSelf) ?>">
    <?php if ($transport === 'hidden'): ?>
      <input type="hidden" name="sid" value="<?= htmlspecialchars(session_id(), ENT_QUOTES, 'UTF-8') ?>">
    <?php endif; ?>
    <label for="name">Update your name:</label><br>
    <input id="name" name="name" type="text" value="<?= htmlspecialchars($name, ENT_QUOTES, 'UTF-8') ?>">
    <button type="submit">Save</button>
  </form>

  <form method="post" action="<?= link_to('/cgi-bin/php/state-demo-2.php', $paramsSelf) ?>" style="margin-top:.75rem">
    <?php if ($transport === 'hidden'): ?>
      <input type="hidden" name="sid" value="<?= htmlspecialchars(session_id(), ENT_QUOTES, 'UTF-8') ?>">
    <?php endif; ?>
    <input type="hidden" name="reset" value="1">
    <button type="submit">Reset Session</button>
  </form>
<?php
if (isset($_POST['reset'])) {
    $_SESSION = [];
    if (session_id() !== '') {
        @session_destroy();
    }
    session_start([
      'cookie_httponly' => true,
      'cookie_secure'   => (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] !== 'off'),
    ]);
    echo '<p><i>Session reset. Reload this page.</i></p>';
}
?>
</body>
</html>

File: public_html/cgi-bin/python/state-demo-2.py
#!/usr/bin/env python3
import os, sys, json, urllib.parse, html, pathlib

SESS_DIR = "/tmp/pycgi_sessions"
pathlib.Path(SESS_DIR).mkdir(parents=True, exist_ok=True)

def read_body():
    length = int(os.environ.get("CONTENT_LENGTH") or 0)
    return sys.stdin.buffer.read(length) if length > 0 else b""

def parse_urlencoded(s):
    return {k: v for k, v in urllib.parse.parse_qs(s, keep_blank_values=True).items()}

def get_sid():
    cookie = os.environ.get("HTTP_COOKIE", "")
    sid = None
    for part in cookie.split(";"):
        k, _, v = part.strip().partition("=")
        if k == "SID":
            sid = v.strip()
            break
    if not sid:
        qs = os.environ.get("QUERY_STRING","")
        sid = urllib.parse.parse_qs(qs).get("sid", [None])[0]
    if not sid and os.environ.get("REQUEST_METHOD") == "POST":
        body = read_body().decode("utf-8", errors="replace")
        sid = parse_urlencoded(body).get("sid", [None])[0]
    return sid

def load_session(sid):
    if not sid: return None
    if not sid.replace("-", "").replace("_", "").isalnum():
        return None
    p = os.path.join(SESS_DIR, f"{sid}.json")
    if not os.path.exists(p): return None
    try:
        with open(p, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return None

sid = get_sid()
sess = load_session(sid)
name = (sess or {}).get("data", {}).get("name", "")

print("Content-Type: text/html; charset=utf-8")
print()
print(f"""<!doctype html>
<html><body style="font-family:system-ui">
  <h1>State Demo — Page 2</h1>
  {"<p><b>Name from session:</b> " + html.escape(name) + "</p>" if name else "<p>No session found.</p>"}
  <p><a href="/cgi-bin/python/state-demo-1.py">Back to Page 1</a></p>
</body></html>""")

